File: core/src/main/java/com/caibowen/prma/core/filter/StrFilter.java
package com.caibowen.prma.core.filter;

import com.caibowen.gplume.context.ClassLoaderInputStreamProvider;
import com.caibowen.gplume.context.InputStreamCallback;
import com.caibowen.gplume.context.InputStreamProviderProxy;
import com.caibowen.gplume.context.InputStreamSupport;
import com.caibowen.gplume.misc.Str;

import javax.inject.Inject;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.stream.StreamSupport;

/**
 * Filter by prefix, suffix or entire string
 *
 *
 * @author BowenCai
 * @since 30-10-2014.
 */
public class StrFilter extends AbstractFilter<String> {

    @Inject private String configPath;
    public String getConfigPath() {
        return configPath;
    }
    public void setConfigPath(String configPath) {
        this.configPath = configPath;
    }

    @Override
    protected int doAccept(String e) {
        return null == ignoreByFullMatch.get(e) ? 1 : -1;
    }

    HashMap<String, Object> ignoreByFullMatch = new HashMap<>(128);

    static Object NA = new Object();

    protected void handle(String buf) {

        /**
         * skip # comment;
         */
        int idx = buf.indexOf('#');
        if (idx != -1 && Str.Utils.isBlank(buf.substring(0, idx)))
            return;

        ignoreByFullMatch.put(buf, NA);
    }

    public void start() {
        InputStreamSupport streamSupport = InputStreamSupport.DEFAULT_SUPPORT;
        streamSupport.withPath(configPath, new InputStreamCallback() {
            @Override
            public void doInStream(InputStream stream) throws Exception {
                BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
                String buf;
                while (null != (buf = reader.readLine())) {
                    if (Str.Utils.notBlank(buf))
                        handle(buf);
                }

                reader.close();
            }
        });
        started = true;
    }

    @Override
    public void stop() {
        ignoreByFullMatch.clear();
        started = false;
    }



}

File: core/src/main/scala/com/caibowen/prma/core/filter/StrFilter.scala
package com.caibowen.prma.core.filter

import java.io.{BufferedReader, InputStream, InputStreamReader}

import com.caibowen.gplume.context.{InputStreamCallback, InputStreamProviderProxy, InputStreamSupport}
import com.caibowen.gplume.misc.Str.Utils
import com.caibowen.gplume.misc.Str.Utils._

import scala.collection.mutable

/**
* @author BowenCai
* @since  01/12/2014.
*/
class StrFilter(configPath: String) extends InputStreamSupport with Filter[String] {

  protected[this] val matchFull = new mutable.HashSet[String]


  @inline
  protected def parse(ptn: String): Unit = {
    matchFull.add(ptn)
  }

  override def start(): Unit = {
    if (started)
      return

    if (isBlank(configPath)) {
      started = true
      return
    }

    if (getStreamProvider == null)
      setStreamProvider(InputStreamProviderProxy.DEFAULT_PROXY)
    withPath(configPath, new InputStreamCallback {
      override def doInStream(stream: InputStream): Unit = {
        val reader = new BufferedReader(new InputStreamReader(stream))
        // how to go FP???
        var buf = reader.readLine();
        while (null != buf) {
          if (notBlank(buf))
            parse(buf);
            buf = reader.readLine();
        }
        reader.close();
      }
    });

    started = true
  }

  @inline
  override def doAccept(str: String): Int = {
    if (matchFull.contains(str))
      1
    else
      -1
  }

  override def stop(): Unit = {
    matchFull.clear()
  }


}
